cpu/esp8266/periph/i2c.c

```diff
@@ -159,13 +163,30 @@ void i2c_init(i2c_t dev)
     DEBUG ("%s: scl=%d sda=%d speed=%d\n", __func__,
            _i2c_bus[dev].scl, _i2c_bus[dev].sda, _i2c_bus[dev].speed);
 
-    /* configure SDA and SCL pin as GPIO in open-drain mode with enabled pull-ups */
-    gpio_init (_i2c_bus[dev].scl, GPIO_OD_PU);
-    gpio_init (_i2c_bus[dev].sda, GPIO_OD_PU);
+    /*
+     * Configure and initialize SDA and SCL pin.
+     * Note: Due to critical timing required by the I2C software
+     * implementation, the ESP8266 GPIOs can not be used directly in GPIO_OD_PU
+     * mode. Instead, the GPIOs are configured in GPIO_IN_PU mode with
+     * open-drain output driver. Signal levels are then realized as following:
+     *
+     * - HIGH: The GPIO is used in the configured GPIO_IN_PU mode. In this
+     *         mode, the output driver is in open-drain mode and pulled-up.
+     * - LOW : The GPIO is temporarily switched to GPIO_OD_PU mode. In this
+     *         mode, the output value 0, which is written during
+     *         initialization, actively drives the output to low.
+     */
+    gpio_init (_i2c_bus[dev].scl, GPIO_IN_PU);
+    gpio_init (_i2c_bus[dev].sda, GPIO_IN_PU);
+    gpio_clear (_i2c_bus[dev].scl);
```

Comment:
I am not an ESP user, but I am wondering why you need to use `gpio_clear` after declare it as input ?

Reply:
That's just for performance reasons. Logic HIGH is realized by disabling the output driver and leaving the output floating and pulled up to HIGH. Logically LOW is realized by activating the output driver and writing a 0 to drive the output active low. To avoid having to write the output every time you change from HIGH to LOW, this is done only once during initialization.